<a href="https://www.kaggle.com/code/abhishridhar/bike-store-sales-analysis?scriptVersionId=248073533" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/bike-store-sample-database/products.csv
/kaggle/input/bike-store-sample-database/brands.csv
/kaggle/input/bike-store-sample-database/categories.csv
/kaggle/input/bike-store-sample-database/order_items.csv
/kaggle/input/bike-store-sample-database/staffs.csv
/kaggle/input/bike-store-sample-database/orders.csv
/kaggle/input/bike-store-sample-database/stores.csv
/kaggle/input/bike-store-sample-database/customers.csv
/kaggle/input/bike-store-sample-database/stocks.csv


In [2]:
import pandas as pd
import numpy as np
import sqlite3 as sql
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

In [3]:
products=pd.read_csv('/kaggle/input/bike-store-sample-database/products.csv')
brands=pd.read_csv('/kaggle/input/bike-store-sample-database/brands.csv')
categories=pd.read_csv('/kaggle/input/bike-store-sample-database/categories.csv')
order_items=pd.read_csv('/kaggle/input/bike-store-sample-database/order_items.csv')
staffs=pd.read_csv('/kaggle/input/bike-store-sample-database/staffs.csv')
orders=pd.read_csv('/kaggle/input/bike-store-sample-database/orders.csv')
stores=pd.read_csv('/kaggle/input/bike-store-sample-database/stores.csv')
customers=pd.read_csv('/kaggle/input/bike-store-sample-database/customers.csv')
stocks=pd.read_csv('/kaggle/input/bike-store-sample-database/stocks.csv')

In [4]:
db=sql.connect('bike_store.db')
type(db)

sqlite3.Connection

In [5]:
print("number of rows in brands: " + str(brands.to_sql('brands', db, if_exists='replace', index=False)))
print("number of rows in categories: " + str(categories.to_sql('categories', db, if_exists='replace', index=False)))
print("number of rows in customers: " + str(customers.to_sql('customers', db, if_exists='replace', index=False)))
print("number of rows in order_items: " + str(order_items.to_sql('order_items', db, if_exists='replace', index=False)))
print("number of rows in orders: " + str(orders.to_sql('orders', db, if_exists='replace', index=False)))
print("number of rows in products: " + str(products.to_sql('products', db, if_exists='replace', index=False)))
print("number of rows in staffs: " + str(staffs.to_sql('staffs', db, if_exists='replace', index=False)))
print("number of rows in stocks: " + str(stocks.to_sql('stocks', db, if_exists='replace', index=False)))
print("number of rows in stores: " + str(stores.to_sql('stores', db, if_exists='replace', index=False)))

number of rows in brands: 9
number of rows in categories: 7
number of rows in customers: 1445
number of rows in order_items: 4722
number of rows in orders: 1615
number of rows in products: 321
number of rows in staffs: 10
number of rows in stocks: 939
number of rows in stores: 3


In [6]:
q1 = '''
SELECT * FROM brands;
'''

df1 = pd.read_sql_query(q1, db) 
df1

brand_id    brand_name
0         1       Electra
1         2          Haro
2         3        Heller
3         4   Pure Cycles
4         5       Ritchey
5         6       Strider
6         7  Sun Bicycles
7         8         Surly
8         9          Trek

In [7]:
q2 = '''
SELECT * FROM categories;
'''

df2 = pd.read_sql_query(q2, db) 
df2

category_id        category_name
0            1    Children Bicycles
1            2     Comfort Bicycles
2            3    Cruisers Bicycles
3            4  Cyclocross Bicycles
4            5       Electric Bikes
5            6       Mountain Bikes
6            7           Road Bikes

In [8]:
q3 = '''
SELECT * FROM customers LIMIT 10;
'''

df3 = pd.read_sql_query(q3, db) 
df3

customer_id  first_name last_name           phone  \
0            1       Debra     Burks            None   
1            2       Kasha      Todd            None   
2            3      Tameka    Fisher            None   
3            4       Daryl    Spence            None   
4            5  Charolette      Rice  (916) 381-6003   
5            6     Lyndsey      Bean            None   
6            7     Latasha      Hays  (716) 986-3359   
7            8   Jacquline    Duncan            None   
8            9    Genoveva   Baldwin            None   
9           10     Pamelia    Newman            None   

                        email                   street             city state  \
0       debra.burks@yahoo.com        9273 Thorne Ave.      Orchard Park    NY   
1        kasha.todd@yahoo.com         910 Vine Street          Campbell    CA   
2       tameka.fisher@aol.com    769C Honey Creek St.     Redondo Beach    CA   
3        daryl.spence@aol.com          988 Pearl Lane         Uniondale    NY   
4     charolette.rice@msn.com           107 River Dr.        Sacramento    CA   
5    lyndsey.bean@hotmail.com           769 West Road          Fairport    NY   
6    latasha.hays@hotmail.com  7014 Manor Station Rd.           Buffalo    NY   
7  jacquline.duncan@yahoo.com            15 Brown St.   Jackson Heights    NY   
8    genoveva.baldwin@msn.com       8550 Spruce Drive   Port Washington    NY   
9    pamelia.newman@gmail.com       476 Chestnut Ave.            Monroe    NY   

   zip_code  
0     14127  
1     95008  
2     90278  
3     11553  
4     95820  
5     14450  
6     14215  
7     11372  
8     11050  
9     10950

In [9]:
q4 = '''
SELECT * FROM order_items LIMIT 10;
'''

df4 = pd.read_sql_query(q4, db) 
df4

order_id  item_id  product_id  quantity  list_price  discount
0         1        1          20         1      599.99      0.20
1         1        2           8         2     1799.99      0.07
2         1        3          10         2     1549.00      0.05
3         1        4          16         2      599.99      0.05
4         1        5           4         1     2899.99      0.20
5         2        1          20         1      599.99      0.07
6         2        2          16         2      599.99      0.05
7         3        1           3         1      999.99      0.05
8         3        2          20         1      599.99      0.05
9         4        1           2         2      749.99      0.10

In [10]:
q5 = '''
SELECT * FROM orders LIMIT 10;
'''

df5 = pd.read_sql_query(q5, db) 
df5

order_id  customer_id  order_status  order_date required_date shipped_date  \
0         1          259             4  2016-01-01    2016-01-03   2016-01-03   
1         2         1212             4  2016-01-01    2016-01-04   2016-01-03   
2         3          523             4  2016-01-02    2016-01-05   2016-01-03   
3         4          175             4  2016-01-03    2016-01-04   2016-01-05   
4         5         1324             4  2016-01-03    2016-01-06   2016-01-06   
5         6           94             4  2016-01-04    2016-01-07   2016-01-05   
6         7          324             4  2016-01-04    2016-01-07   2016-01-05   
7         8         1204             4  2016-01-04    2016-01-05   2016-01-05   
8         9           60             4  2016-01-05    2016-01-08   2016-01-08   
9        10          442             4  2016-01-05    2016-01-06   2016-01-06   

   store_id  staff_id  
0         1         2  
1         2         6  
2         2         7  
3         1         3  
4         2         6  
5         2         6  
6         2         6  
7         2         7  
8         1         2  
9         2         6

In [11]:
q6 = '''
SELECT * FROM products LIMIT 10;
'''

df6 = pd.read_sql_query(q6, db) 
df6

product_id                           product_name  brand_id  category_id  \
0           1                        Trek 820 - 2016         9            6   
1           2     Ritchey Timberwolf Frameset - 2016         5            6   
2           3        Surly Wednesday Frameset - 2016         8            6   
3           4               Trek Fuel EX 8 29 - 2016         9            6   
4           5           Heller Shagamaw Frame - 2016         3            6   
5           6  Surly Ice Cream Truck Frameset - 2016         8            6   
6           7               Trek Slash 8 27.5 - 2016         9            6   
7           8  Trek Remedy 29 Carbon Frameset - 2016         9            6   
8           9                   Trek Conduit+ - 2016         9            5   
9          10                 Surly Straggler - 2016         8            4   

   model_year  list_price  
0        2016      379.99  
1        2016      749.99  
2        2016      999.99  
3        2016     2899.99  
4        2016     1320.99  
5        2016      469.99  
6        2016     3999.99  
7        2016     1799.99  
8        2016     2999.99  
9        2016     1549.00

In [12]:
q7 = '''
SELECT * FROM staffs;
'''

df7 = pd.read_sql_query(q7, db) 
df7

/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()


staff_id  first_name last_name                          email  \
0         1     Fabiola   Jackson     fabiola.jackson@bikes.shop   
1         2      Mireya  Copeland     mireya.copeland@bikes.shop   
2         3       Genna   Serrano       genna.serrano@bikes.shop   
3         4      Virgie   Wiggins      virgie.wiggins@bikes.shop   
4         5    Jannette     David      jannette.david@bikes.shop   
5         6   Marcelene     Boyer     marcelene.boyer@bikes.shop   
6         7      Venita    Daniel       venita.daniel@bikes.shop   
7         8        Kali    Vargas         kali.vargas@bikes.shop   
8         9       Layla   Terrell       layla.terrell@bikes.shop   
9        10  Bernardine   Houston  bernardine.houston@bikes.shop   

            phone  active  store_id  manager_id  
0  (831) 555-5554       1         1         NaN  
1  (831) 555-5555       1         1         1.0  
2  (831) 555-5556       1         1         2.0  
3  (831) 555-5557       1         1         2.0  
4  (516) 379-4444       1         2         1.0  
5  (516) 379-4445       1         2         5.0  
6  (516) 379-4446       1         2         5.0  
7  (972) 530-5555       1         3         1.0  
8  (972) 530-5556       1         3         7.0  
9  (972) 530-5557       1         3         7.0

In [13]:
q8 = '''
SELECT * FROM stocks LIMIT 10;
'''

df8 = pd.read_sql_query(q8, db) 
df8

store_id  product_id  quantity
0         1           1        27
1         1           2         5
2         1           3         6
3         1           4        23
4         1           5        22
5         1           6         0
6         1           7         8
7         1           8         0
8         1           9        11
9         1          10        15

In [14]:
q9 = '''
SELECT * FROM stores;
'''

df9 = pd.read_sql_query(q9, db) 
df9

store_id        store_name           phone                 email  \
0         1  Santa Cruz Bikes  (831) 476-4321  santacruz@bikes.shop   
1         2     Baldwin Bikes  (516) 379-8888    baldwin@bikes.shop   
2         3     Rowlett Bikes  (972) 530-5555    rowlett@bikes.shop   

                street        city state  zip_code  
0   3700 Portola Drive  Santa Cruz    CA     95060  
1   4200 Chestnut Lane     Baldwin    NY     11432  
2  8000 Fairway Avenue     Rowlett    TX     75088

# **Customer Behavior & Demographics Analysis**


In [15]:
q11 = '''
SELECT str.state,
'$' || ROUND(SUM(ort.quantity*(ort.list_price*(1-ort.discount))) / COUNT(DISTINCT ort.order_id),2)
AS average_revenue_per_order
FROM stores str 
LEFT JOIN orders o 
ON o.store_id=str.store_id 
LEFT JOIN order_items ort
ON ort.order_id=o.order_id GROUP BY str.state ORDER BY average_revenue_per_order DESC;
'''

df11 = pd.read_sql_query(q11, db) 
df11

state average_revenue_per_order
0    TX                  $4985.87
1    NY                  $4771.96
2    CA                  $4614.43

Identifies the average revenue a customer would make per order (total revenue/total orders) grouped by state.
As we can see, Texans tend to spend the most per order, so online services could be informed to focus on Texas markets.

In [16]:
q12 = '''
SELECT status, COUNT(status) AS 'Amount'  FROM
(
SELECT cu.first_name,cu.last_name,
CASE WHEN COUNT(DISTINCT o.order_id)>1 THEN 'returning'
ELSE 'new'
END AS 'status'
FROM customers cu 
LEFT JOIN orders o ON 
o.customer_id=cu.customer_id
GROUP BY cu.customer_id
)
GROUP BY status;
'''

df12 = pd.read_sql_query(q12, db) 
df12

status  Amount
0        new    1314
1  returning     131

Identifies how many customers are new and how many are returning via a subquery for all three stores. 

In [17]:
qx = '''
SELECT 
  store_name,
  COUNT(CASE WHEN status = 'returning' THEN 1 END) AS returning_customers,
  COUNT(CASE WHEN status = 'new' THEN 1 END) AS new_customers,
  ROUND(
    100.0 * COUNT(CASE WHEN status = 'returning' THEN 1 END) 
    / COUNT(*), 
    2
  ) || '%' AS returning_percentage
FROM (
  SELECT 
    cu.customer_id,
    s.store_name,
    CASE 
      WHEN COUNT(DISTINCT o.order_id) > 1 THEN 'returning'
      ELSE 'new'
    END AS status
  FROM customers cu
  LEFT JOIN orders o ON o.customer_id = cu.customer_id
  LEFT JOIN stores s ON s.store_id = o.store_id
  GROUP BY cu.customer_id, s.store_name
) customer_status_by_store
GROUP BY store_name ORDER BY 
ROUND(
    100.0 * COUNT(CASE WHEN status = 'returning' THEN 1 END) 
    / COUNT(*), 
    2
  ) DESC
'''

dfx = pd.read_sql_query(qx, db) 
dfx

store_name  returning_customers  new_customers returning_percentage
0  Santa Cruz Bikes                   52            232               18.31%
1     Rowlett Bikes                   19            123               13.38%
2     Baldwin Bikes                   60            959                5.89%

Although Baldwin Bikes have had the most returning customers, Santa Cruz Bikes has the largest retention rate of customers, with 18.31% of all of its customers returning.

In [18]:
q13 = '''
SELECT c.state, COUNT(c.customer_id) AS Number_of_people 
FROM customers c 
GROUP BY c.state 
ORDER BY Number_of_people DESC; 
'''

df13 = pd.read_sql_query(q13, db) 
df13

state  Number_of_people
0    NY              1019
1    CA               284
2    TX               142

The majority of people in this database are concentrated in New York

In [19]:
q14 = '''
SELECT c.zip_code, c.state, COUNT(c.customer_id) AS Number_of_people 
FROM customers c
GROUP BY c.zip_code 
ORDER BY Number_of_people DESC LIMIT 20; 
'''

df14 = pd.read_sql_query(q14, db) 
df14

zip_code state  Number_of_people
0      10550    NY                20
1      12020    NY                17
2      10583    NY                17
3      14424    NY                14
4      75604    TX                13
5      11001    NY                13
6      10562    NY                13
7      91387    CA                12
8      76901    TX                12
9      14580    NY                12
10     11787    NY                12
11     11779    NY                12
12     11566    NY                12
13     11418    NY                12
14     11414    NY                12
15     11104    NY                12
16     11102    NY                12
17     92806    CA                11
18     92307    CA                11
19     91784    CA                11

Group the number of people in the database by zip code.

# **Sales & Revenue Analysis**


In [20]:
q15 = '''
SELECT b.brand_name, '$' ||
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000),1) || ' thousand'
AS revenue
FROM brands b 
LEFT JOIN products p 
ON p.brand_id=b.brand_id
LEFT JOIN order_items ort 
ON ort.product_id=p.product_id 
GROUP BY b.brand_name 
ORDER BY 
ROUND(((SUM(ort.quantity*(ort.list_price-(ort.discount*ort.list_price))))/1000),2) DESC;



'''

df15 = pd.read_sql_query(q15, db) 
df15

brand_name           revenue
0          Trek  $4602.8 thousand
1       Electra  $1205.3 thousand
2         Surly   $949.5 thousand
3  Sun Bicycles   $342.0 thousand
4          Haro   $185.4 thousand
5        Heller   $171.5 thousand
6   Pure Cycles   $149.5 thousand
7       Ritchey    $78.9 thousand
8       Strider     $4.3 thousand

The Trek bike brand is doing the best in terms of generating revenue across all three stores.

In [21]:
qy = '''

WITH brand_revenue AS (
  SELECT 
    s.store_name AS store,
    b.brand_name,
    SUM(ort.quantity * (ort.list_price * (1 - ort.discount))) AS total_revenue
  FROM brands b
  JOIN products p ON p.brand_id = b.brand_id
  JOIN order_items ort ON ort.product_id = p.product_id
  JOIN orders o ON o.order_id = ort.order_id
  JOIN stores s ON s.store_id = o.store_id
  GROUP BY s.store_name, b.brand_name
)
SELECT store, brand_name,
  '$' || ROUND(total_revenue / 1000, 1) || ' thousand' AS revenue
FROM brand_revenue br
WHERE total_revenue = (
  SELECT MAX(total_revenue)
  FROM brand_revenue
  WHERE store = br.store
)
ORDER BY store;
'''

dfy = pd.read_sql_query(qy, db) 
dfy

store brand_name           revenue
0     Baldwin Bikes       Trek  $3124.3 thousand
1     Rowlett Bikes       Trek   $552.7 thousand
2  Santa Cruz Bikes       Trek   $925.8 thousand

Identifies the best-selling brand by store.

In [22]:
q16= '''
SELECT p.product_name, SUM(ort.quantity) AS Total_quantity_ordered, '$' ||
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))),2) AS total_revenue
FROM products p
LEFT JOIN order_items ort ON ort.product_id=p.product_id
GROUP BY p.product_name ORDER BY SUM(ort.quantity) DESC,
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))),2) DESC LIMIT 10;

'''
df16 = pd.read_sql_query(q16, db) 
df16

product_name  Total_quantity_ordered  \
0             Electra Cruiser 1 (24-Inch) - 2016                     296   
1           Electra Townie Original 7D EQ - 2016                     290   
2             Electra Townie Original 21D - 2016                     289   
3  Electra Girl's Hawaii 1 (16-inch) - 2015/2016                     269   
4          Surly Ice Cream Truck Frameset - 2016                     167   
5                       Trek Slash 8 27.5 - 2016                     154   
6  Electra Girl's Hawaii 1 (20-inch) - 2015/2016                     154   
7                    Surly Straggler 650b - 2016                     151   
8         Electra Townie Original 7D - 2015/2016                     148   
9                         Surly Straggler - 2016                     147   

  total_revenue  
0     $72071.13  
1    $155169.41  
2    $143393.39  
3     $64514.11  
4      $70371.6  
5    $555558.61  
6     $41011.63  
7    $226765.55  
8     $66568.67  
9    $203507.62

The Trek Slash 8 27.5 has made the most total revenue overall with 555k in sales across all 3 stores, but it doesn't make top the list in terms of quantity ordered, with the Electra Cruiser 1 (24-inch) topping the list with 296 sales and 72k in total revenue.

In [23]:
q17 = '''
SELECT cat.category_name, 
'$' ||
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000),2) || ' thousand'
AS total_revenue
FROM categories cat 
LEFT JOIN products p 
ON p.category_id=cat.category_id
LEFT JOIN order_items ort ON ort.product_id=p.product_id
GROUP BY cat.category_name
ORDER BY ((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000) DESC;



'''

df17 = pd.read_sql_query(q17, db) 
df17

category_name      total_revenue
0       Mountain Bikes  $2715.08 thousand
1           Road Bikes   $1665.1 thousand
2    Cruisers Bicycles   $995.03 thousand
3       Electric Bikes   $916.68 thousand
4  Cyclocross Bicycles   $711.01 thousand
5     Comfort Bicycles   $394.02 thousand
6    Children Bicycles   $292.19 thousand

Mountain bikes seem to be the best-selling category of bikes, with 2.7 million dollars in total revenue across all brands.

In [24]:
q18 = '''
SELECT staffs.first_name, staffs.last_name, s.store_name, 
'$' ||
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000),2) || ' thousand' 
AS revenue
FROM staffs
LEFT JOIN orders o 
ON o.staff_id=staffs.staff_id
LEFT JOIN order_items ort 
ON ort.order_id=o.order_id
LEFT JOIN stores s 
ON s.store_id=staffs.store_id
GROUP BY staffs.staff_id 
ORDER BY 
((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))) DESC;



'''

df18 = pd.read_sql_query(q18, db) 
df18

first_name last_name        store_name            revenue
0   Marcelene     Boyer     Baldwin Bikes  $2624.12 thousand
1      Venita    Daniel     Baldwin Bikes  $2591.63 thousand
2       Genna   Serrano  Santa Cruz Bikes   $853.29 thousand
3      Mireya  Copeland  Santa Cruz Bikes   $752.54 thousand
4        Kali    Vargas     Rowlett Bikes   $463.92 thousand
5       Layla   Terrell     Rowlett Bikes   $403.62 thousand
6  Bernardine   Houston     Rowlett Bikes               None
7    Jannette     David     Baldwin Bikes               None
8      Virgie   Wiggins  Santa Cruz Bikes               None
9     Fabiola   Jackson  Santa Cruz Bikes               None

Marcelene Boyer, followed by Venita Daniel, has generated the most revenue across all three stores. We can create a subquery to show the most productive employee for each store.

In [25]:
q19 = '''
SELECT store_name, first_name, last_name, MAX(revenue) as revenue FROM
(
SELECT staffs.first_name, staffs.last_name, s.store_name, 
'$' ||
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000),2) || ' thousand' 
AS revenue
FROM staffs
LEFT JOIN orders o 
ON o.staff_id=staffs.staff_id
LEFT JOIN order_items ort 
ON ort.order_id=o.order_id
LEFT JOIN stores s 
ON s.store_id=staffs.store_id
GROUP BY staffs.staff_id 
ORDER BY 
((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))) DESC
)
GROUP BY store_name;



'''

df19 = pd.read_sql_query(q19, db) 
df19

store_name first_name last_name            revenue
0     Baldwin Bikes  Marcelene     Boyer  $2624.12 thousand
1     Rowlett Bikes       Kali    Vargas   $463.92 thousand
2  Santa Cruz Bikes      Genna   Serrano   $853.29 thousand

This shows the most productive employee (revenue-generating) for each store.

In [26]:
q20 = '''
SELECT s.store_name,
'$' ||
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000),2) || ' thousand' 
AS revenue
FROM stores s
LEFT JOIN orders o 
ON o.store_id=s.store_id
LEFT JOIN order_items ort 
ON ort.order_id=o.order_Id
GROUP BY s.store_name 
ORDER BY 
ROUND(((SUM(ort.quantity*(ort.list_price*(1-ort.discount))))/1000),2) DESC;



'''

df20 = pd.read_sql_query(q20, db) 
df20

store_name            revenue
0     Baldwin Bikes  $5215.75 thousand
1  Santa Cruz Bikes  $1605.82 thousand
2     Rowlett Bikes   $867.54 thousand

Baldwin Bikes generates the highest revenue overall with $5.2 million. Given that the majority of customers in this database are from NY and Baldwin is in NY, this result is predictable.

# **Inventory & Operations Analysis**


In [27]:
q21= '''

SELECT p.product_name 
AS products_out_of_stock 
FROM products p
LEFT JOIN stocks s
ON s.product_id=p.product_id
LEFT JOIN stores st
ON st.store_id=s.store_id
WHERE s.quantity=0 AND store_name='Baldwin Bikes'


'''

df21 = pd.read_sql_query(q21, db) 
df21

products_out_of_stock
0  Electra Girl's Hawaii 1 (16-inch) - 2015/2016
1                         Trek Remedy 9.8 - 2017
2   Trek Precaliber 24 (21-Speed) - Girls - 2017
3                         Trek CrossRip 1 - 2018
4                Trek Domane SLR Frameset - 2018
5                   Trek Domane SL 6 Disc - 2018
6              Electra Townie Go! 8i - 2017/2018
7              Electra Townie Commute Go! - 2018
8              Electra Townie Commute Go! - 2018
9             Electra Townie Original 21D - 2018

This represents the products that are out of stock for Baldwin's bikes.

By altering the store name in the WHERE clause, we can see the out-of-stock products for each store.

In [28]:
q21a= '''
SELECT store_name, COUNT(products_out_of_stock) AS Amount_out_of_stock FROM
(
SELECT st.store_name,p.product_name 
AS products_out_of_stock 
FROM products p
LEFT JOIN stocks s
ON s.product_id=p.product_id
LEFT JOIN stores st
ON st.store_id=s.store_id
WHERE s.quantity=0
)
GROUP BY
store_name
ORDER BY
Amount_out_of_stock DESC;



'''

df21a = pd.read_sql_query(q21a, db) 
df21a

store_name  Amount_out_of_stock
0  Santa Cruz Bikes                   10
1     Baldwin Bikes                   10
2     Rowlett Bikes                    5

In [29]:
q22= '''

SELECT p.product_name 
AS products_in_low_stock
FROM products p
LEFT JOIN stocks s
ON s.product_id=p.product_id
LEFT JOIN stores st
ON st.store_id=s.store_id
WHERE s.quantity<=5 AND store_name='Santa Cruz Bikes'


'''

df22 = pd.read_sql_query(q22, db) 
df22

products_in_low_stock
0                Ritchey Timberwolf Frameset - 2016
1             Surly Ice Cream Truck Frameset - 2016
2             Trek Remedy 29 Carbon Frameset - 2016
3                             Electra Moto 1 - 2016
4              Electra Townie Original 7D EQ - 2016
..                                              ...
57                          Trek Superfly 20 - 2018
58    Electra Straight 8 1 (16-inch) - Boy's - 2018
59                   Electra Superbolt 1 20" - 2018
60  Electra Tiger Shark 3i (20-inch) - Boys' - 2018
61                 Electra Townie Original 1 - 2018

[62 rows x 1 columns]

This represents the products in low stock (less than 5) in Santa Cruz Bikes.

Again, the store name in the WHERE clause can be adjusted to show each store's data.

In [30]:
q22a= '''
SELECT store_name, COUNT(products_in_low_stock) AS Amount_low_in_stock FROM
(
SELECT st.store_name, product_name 
AS products_in_low_stock
FROM products p
LEFT JOIN stocks s
ON s.product_id=p.product_id
LEFT JOIN stores st
ON st.store_id=s.store_id
WHERE s.quantity<=5
)
GROUP BY store_name ORDER BY Amount_low_in_stock DESC;


'''

df22a = pd.read_sql_query(q22a, db) 
df22a

store_name  Amount_low_in_stock
0     Baldwin Bikes                   66
1  Santa Cruz Bikes                   62
2     Rowlett Bikes                   62

In [31]:
q23= '''

SELECT 
strftime('%m', o.order_date) AS month, '$' ||
ROUND(SUM(ort.quantity * ort.list_price * (1 - ort.discount)), 2) AS revenue
FROM orders o
LEFT JOIN order_items ort 
ON ort.order_id = o.order_id
WHERE strftime('%Y', o.order_date) = '2016'
GROUP BY strftime('%m', o.order_date) 
ORDER BY
ROUND(SUM(ort.quantity * ort.list_price * (1 - ort.discount)), 2) DESC;


'''

df23 = pd.read_sql_query(q23, db) 
df23

month     revenue
0     09  $273091.61
1     08  $225657.38
2     01  $215146.42
3     10  $212078.08
4     06  $210562.12
5     05  $205270.01
6     12  $199829.98
7     07  $199556.81
8     11  $182329.41
9     03  $180600.33
10    04  $167144.05
11    02  $156112.32

This shows the total revenue for all bike stores grouped by every month in 2016. For all stores cumulative, the best month was September. Month can be adjusted with the WHERE clause.

To show each store, we join the 'stores' table and add a WHERE clause to define the specific store we want data.

In [32]:
q24= '''

SELECT 
strftime('%Y', o.order_date) AS year, '$' ||
ROUND(SUM(ort.quantity * ort.list_price * (1 - ort.discount)), 2) AS revenue
FROM orders o
LEFT JOIN order_items ort 
ON ort.order_id = o.order_id
GROUP BY strftime('%Y', o.order_date) 
ORDER BY year DESC;


'''

df24 = pd.read_sql_query(q24, db) 
df24

year      revenue
0  2018  $1814529.79
1  2017  $3447208.24
2  2016  $2427378.53

We find that the best-performing year for all three stores cumulative was 2017, with 3.4 million combined from all stores.

# **Marketing Analytics**

In [33]:


q25= '''


SELECT cu.first_name,cu.last_name
FROM customers cu 
LEFT JOIN orders o 
ON o.customer_id=cu.customer_id
LEFT JOIN stores s
ON s.store_id=o.store_id
WHERE s.store_name='Baldwin Bikes'
GROUP BY cu.customer_id 
HAVING COUNT(DISTINCT o.order_id)>=3



'''

df25 = pd.read_sql_query(q25, db) 
df25

first_name last_name
0       Debra     Burks
1       Daryl    Spence
2     Lyndsey      Bean
3     Latasha      Hays
4   Jacquline    Duncan
5    Genoveva   Baldwin
6     Pamelia    Newman
7     Deshawn   Mendoza
8       Robby     Sykes
9      Linnie    Branch
10     Emmitt   Sanchez
11      Caren  Stephens
12  Georgetta    Hardin
13   Lizzette     Stein

This shows loyal customers (Ordering 3 or more times) from Baldwin Bikes. 

The WHERE clause in the query can be used to change the store name and find loyal customers from each store.

In [34]:
q26='''
SELECT cu.first_name, cu.last_name,s.state, s.zip_code, '$' ||
ROUND(SUM(ort.quantity*(ort.list_price-(ort.discount*ort.list_price))),2) AS total_spent
FROM customers cu 
LEFT JOIN orders o 
ON o.customer_id=cu.customer_id
LEFT JOIN order_items ort 
ON ort.order_id=o.order_id 
LEFT JOIN stores s 
ON s.store_id=o.store_id
WHERE s.store_name='Santa Cruz Bikes'
GROUP BY cu.customer_id
ORDER BY ROUND(SUM(ort.quantity*(ort.list_price-(ort.discount*ort.list_price))),2) 
DESC LIMIT 10;
'''
df26=pd.read_sql_query(q26,db)
df26

first_name  last_name state  zip_code total_spent
0     Tameka     Fisher    CA     95060   $24051.53
1     Brigid      Sharp    CA     95060   $20648.95
2  Williemae   Holloway    CA     95060   $19756.64
3      Kasha       Todd    CA     95060   $19329.08
4    Deloris      Burke    CA     95060    $19043.9
5     Lizzie     Joyner    CA     95060   $18995.01
6     Marvin    Mullins    CA     95060   $18856.88
7    Araceli     Golden    CA     95060   $18037.29
8     Carola  Rodriquez    CA     95060    $17861.2
9       Paul     Lester    CA     95060   $17580.72

This shows the top-spending customers for Santa Cruz Bikes. We can look at each store's data by changing the store in the WHERE clause.